File: java_pro/src/main/java/jv/com/tencent/cain/design_pattern/observer/Subject.java
package jv.com.tencent.cain.design_pattern.observer;

/**
 * @author : jiangyu
 * @date : 2020/3/3
 * @desc : 观察者设计模式-通知者
 */
public interface Subject {
    /**
     * 添加观察者
     *
     * @param observer
     */
    void addObserver(Observer observer);

    /**
     * 移除观察者
     *
     * @param observer
     */
    void removeObserver(Observer observer);

    /**
     * 全部移除
     */
    void removeAll();

    /**
     * 通知所有观察者
     *
     * @param data
     */
    void notifyAllObserver(Object data);

    /**
     * 通知某个观察者
     *
     * @param observer
     * @param data
     */
    void notify(Observer observer, Object data);
}

File: java_pro/src/main/kotlin/kt/com/tencent/cain/desigin_pattern/observer/Subject.kt
package kt.com.tencent.cain.desigin_pattern.observer

/**
 * @author : jiangyu
 * @date   : 2020/3/3
 * @desc   : 观察者设计模式-通知者
 */
interface Subject {
    /**
     * 添加观察者
     */
    fun addObserver(observer: Observer)

    /**
     * 移除观察者
     */
    fun remove(observer: Observer)

    /**
     * 全部移除
     */
    fun removeAll()

    /**
     * 通知所有观察者
     */
    fun notifyAll(data: Any)

    /**
     * 通知某个观察者
     */
    fun notify(observer: Observer, data: Any)
}